presto-main/src/main/java/com/facebook/presto/sql/gen/PageProcessorCompiler.java

```diff
@@ -88,98 +96,48 @@ private void generateProcessMethod(ClassDefinition classDefinition, RowExpressio
+            builder.add(blockBuilder);
+        }
+        List<Variable> blockBuilders = builder.build();
 
+        // projection body
+        Variable position = scope.declareVariable(int.class, "position");
+        ByteCodeBlock project = new ByteCodeBlock().append(pageBuilder.invoke("declarePosition", void.class));
         for (int projectionIndex = 0; projectionIndex < projections.size(); projectionIndex++) {
-            List<Integer> inputChannels = getInputChannels(projections.get(projectionIndex));
-
-            filterBlock.ifTrue()
-                    .append(thisVariable)
-                    .append(session)
-                    .append(pushBlockVariables(scope, inputChannels))
-                    .getVariable(position);
-
-            filterBlock.ifTrue()
-                    .comment("pageBuilder.getBlockBuilder(%d)", projectionIndex)
-                    .append(pageBuilder)
-                    .push(projectionIndex)
-                    .invokeVirtual(PageBuilder.class, "getBlockBuilder", BlockBuilder.class, int.class);
-
-            filterBlock.ifTrue()
-                    .comment("project_%d(session, block_%s, position, blockBuilder)", projectionIndex, inputChannels)
-                    .invokeVirtual(classDefinition.getType(),
-                            "project_" + projectionIndex,
-                            type(void.class),
-                            ImmutableList.<ParameterizedType>builder()
-                                    .add(type(ConnectorSession.class))
-                                    .addAll(nCopies(inputChannels.size(), type(Block.class)))
-                                    .add(type(int.class))
-                                    .add(type(BlockBuilder.class))
-                                    .build());
+            RowExpression projection = projections.get(projectionIndex);
+            project.append(invokeProject(thisVariable, session, expressionInputBlocks.get(projection), position, blockBuilders.get(projectionIndex), projectionMethods.get(projectionIndex)));
         }
+        LabelNode done = new LabelNode("done");
 
-        loopBody.append(filterBlock);
+        // for loop loop body
+        ForLoop loop = new ForLoop()
+                .initialize(position.set(start))
+                .condition(lessThan(position, end))
+                .update(position.set(add(position, constantInt(1))))
+                .body(new ByteCodeBlock()
+                        .append(new IfStatement()
+                                .condition(pageBuilder.invoke("isFull", boolean.class))
+                                .ifTrue(jump(done)))
+                        .append(new IfStatement()
+                                .condition(invokeFilter(thisVariable, session, expressionInputBlocks.get(filter), position))
+                                .ifTrue(project)));
 
         method.getBody()
                 .append(loop)
```

Comment:
in the lines below this line you should be able to do
``
method.getBody()
        .append(loop)
        .visitLabel(done)
        .append(position.ret());
```